File: 2025/day_8/day_8.py
import os
import math

file_path = os.path.join(os.path.dirname(__file__), "input_day_8.txt")
with open(file_path, "r") as file:
    input_ = file.read()


# Custom Disjoint Set Union implementation
class DSU:
    def __init__(self, n):
        self.parent = list(range(n))
        self.size = [1] * n
        self.max_size = 1

    def find(self, i):
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    def merge(self, i, j):
        irep = self.find(i)
        jrep = self.find(j)
        if irep != jrep:
            self.parent[irep] = jrep
            self.size[jrep] += self.size[irep]
            self.size[irep] = 1
            self.max_size = max(self.max_size, self.size[jrep])


# Returns the distance between two 3D points
def dist(c1, c2) -> float:
    return ((c1[0] - c2[0]) ** 2 + (c1[1] - c2[1]) ** 2 + (c1[2] - c2[2]) ** 2) ** 0.5


coords = [tuple(int(x) for x in line.split(",")) for line in input_.split("\n")]
edges = [(i, j) for i in range(len(coords)) for j in range(i + 1, len(coords))]

dsu = DSU(len(coords))
for i, edge in enumerate(sorted(edges, key=lambda e: dist(coords[e[0]], coords[e[1]]))):
    dsu.merge(edge[0], edge[1])
    if i == 999:
        print("Part 1 Answer:", math.prod(sorted(dsu.size, reverse=True)[:3]))
    if dsu.max_size == len(coords):
        print("Part 2 Answer:", coords[edge[0]][0] * coords[edge[1]][0])
        break

File: 2025/day_8/day_8.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

class DSU {
 private:
  std::vector<int> parent;

 public:
  std::vector<int> sizes;
  int max_size = 1;

  DSU(int n) {
    parent = std::vector<int>(n);
    for (int i = 0; i < parent.size(); i++) parent[i] = i;
    sizes = std::vector<int>(n, 1);
  }

  int find(int i) {
    if (parent[i] == i) return i;
    parent[i] = find(parent[i]);
    return parent[i];
  }

  void merge(int i, int j) {
    int i_rep = find(i);
    int j_rep = find(j);
    if (i_rep != j_rep) {
      parent[i_rep] = j_rep;
      sizes[j_rep] += sizes[i_rep];
      sizes[i_rep] = 1;
      max_size = std::max(max_size, sizes[j_rep]);
    }
  }
};

struct Coord {
  int x, y, z;
};

struct Edge {
  int i, j;
  long long length;

  Edge(int i_, int j_, const std::vector<Coord>& coords) : i(i_), j(j_) {
    const auto& a = coords[i];
    const auto& b = coords[j];
    length = 1LL * (a.x - b.x) * (a.x - b.x) + 1LL * (a.y - b.y) * (a.y - b.y) +
             1LL * (a.z - b.z) * (a.z - b.z);
  }
};

int main() {
  std::ifstream file("input_day_8.txt");
  if (!file.is_open()) {
    std::cerr << "Failed to open input\n";
    return 1;
  }

  std::vector<Coord> coords;
  int a, b, c;
  char ignore;

  while (file >> a >> ignore >> b >> ignore >> c) {
    coords.emplace_back(Coord{a, b, c});
  }

  std::vector<Edge> edges;
  edges.reserve(coords.size() * (coords.size() - 1) / 2);

  for (int i = 0; i < coords.size(); i++) {
    for (int j = i + 1; j < coords.size(); j++) {
      edges.emplace_back(Edge(i, j, coords));
    }
  }

  std::sort(edges.begin(), edges.end(),
            [](const auto& a, const auto& b) { return a.length < b.length; });

  DSU dsu = DSU(coords.size());

  for (int k = 0; k < edges.size(); k++) {
    dsu.merge(edges[k].i, edges[k].j);
    if (k == 999) {
      int a = 0, b = 0, c = 0;
      for (int x : dsu.sizes) {
        if (x > a) {
          c = b;
          b = a;
          a = x;
        } else if (x > b) {
          c = b;
          b = x;
        } else if (x > c) {
          c = x;
        }
      }
      std::cout << "Part 1 Answer: " << a * b * c << "\n";
    }
    if (dsu.max_size == coords.size()) {
      std::cout << "Part 2 Answer: "
                << 1LL * coords[edges[k].i].x * coords[edges[k].j].x << "\n";
      break;
    }
  }
}